internal/cephfs/nodeserver.go

```diff
@@ -240,9 +262,100 @@ func (*NodeServer) mount(
 		return status.Error(codes.Internal, err.Error())
 	}
 
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		unmountErr := mounter.UnmountAll(ctx, stagingTargetPath)
+		if unmountErr != nil {
+			log.ErrorLog(ctx, "failed to clean up mounts in rollback procedure: %v", unmountErr)
+		}
+	}()
+
+	if volOptions.BackingSnapshot {
+		snapshotRoot, err := getBackingSnapshotRoot(ctx, volOptions, stagingTargetPath)
+		if err != nil {
+			return err
+		}
+
+		absoluteSnapshotRoot := path.Join(stagingTargetPath, snapshotRoot)
+		err = mounter.BindMount(
+			ctx,
+			absoluteSnapshotRoot,
+			stagingTargetPath,
+			true,
+			[]string{"bind", "_netdev"},
+		)
+
+		if err != nil {
+			log.ErrorLog(ctx,
+				"failed to bind mount snapshot root %s: %v", absoluteSnapshotRoot, err)
+
+			return status.Error(codes.Internal, err.Error())
+		}
+	}
+
 	return nil
 }
 
+func getBackingSnapshotRoot(
+	ctx context.Context,
+	volOptions *store.VolumeOptions,
+	stagingTargetPath string,
+) (string, error) {
+	if volOptions.ProvisionVolume {
+		// Provisioned snapshot-backed volumes should have their BackingSnapshotRoot
+		// already populated.
+		return volOptions.BackingSnapshotRoot, nil
+	}
+
+	// For pre-provisioned snapshot-backed volumes we need to traverse
+	// <RootPath>/.snap in order to find out the correct snapshot directory.
+
+	snapshotsBase := path.Join(stagingTargetPath, ".snap")
+
+	dir, err := os.Open(snapshotsBase)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to open %s when searching for snapshot root: %v", snapshotsBase, err)
+
+		return "", status.Errorf(codes.Internal, err.Error())
+	}
+
+	contents, err := dir.Readdirnames(0)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to read %s when searching for snapshot root: %v", snapshotsBase, err)
+
+		return "", status.Errorf(codes.Internal, err.Error())
+	}
+
+	var (
+		found           bool
+		snapshotDirName string
+	)
+
+	for i := range contents {
```

Comment:
add an example how what content will look like. this helps to understand this logic for someone who reads the code later.